### ISSUE TITLE
Change the build system

### ISSUE THREAD

**Issue by tathanhdinh** [jbaublitz/knock-out#5]
Hi @jbaublitz 

Thank you for taking time to review few recent trivial PRs of mine and giving constructive feedbacks.

This issue is rather a discussion in which I hope receive some feedback from you.  I'm still learning to implement Linux kernel modules in Rust, so please correct me if I was wrong at some points.

### Problem 1
I'm working on a [forked branch](https://github.com/tathanhdinh/knock-out/tree/system_log) where I try to add an explicit error handling for `parrot`, something like that:

```rust
enum ParrorError {
    CharDevAdd,
    CharDevRegionAlloc,
}

type Result<T> = result::Result<T, ParrorError>;

impl fmt::Display for ParrorError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        use ParrorError::*;
        match self {
            CharDevAdd => {
                write!(f, "{}", "Failed to add char dev\0")
            },

            CharDevRegionAlloc => {
                write!(f, "{}", "Failed to allocate char device region\0")
            }
        }
    }
}

#[derive(Default)]
struct KernLog;

impl fmt::Write for KernLog {
    fn write_str(&mut self, s: &str) -> result::Result<(), fmt::Error> {
        unsafe { printk(s.as_ptr()); }
        Ok(())
    }
}

static mut KERNEL_LOG: KernLog = KernLog;
```

so I can use it later, for example

```rust
#[no_mangle]
#[link_section = ".text"]
pub extern "C" fn parrot_init_module() -> i32 {
    let parrot_safe = match ParrotSafe::new() {
        Ok(ps) => ps,
        Err(e) => {
            unsafe { write!(KERNEL_LOG, "{}", e).unwrap() };
            return -1;
        }
    };
    unsafe { GLOBAL_STATE = Some(parrot_safe) };
    0
}

#[no_mangle]
#[link_section = ".text"]
pub extern "C" fn parrot_cleanup_module() {
    unsafe {
        match GLOBAL_STATE {
            Some(ref mut ps) => {
                match ps.cleanup() {
                    Ok(_) => (),
                    Err(e) => {
                        write!(KERNEL_LOG, "{}", e).unwrap();
                    }
                }
            }
            None => (),
        }
    }
}
```

But then I discovered that the current object code generation `--emit obj=src/rust.o` no longer work, that is because there is a need of multiple objects generation (is this true?). 

For example, on the current parrot, the symbol table of `src/rust.o`:

```bash
                 U alloc_chrdev_region
                 U cdev_add
                 U cdev_del
                 U cdev_init
                 U cdev_ptr
00000000000000d0 T cleanup_module
                 U copy_to_user_ffi
                 U fops_ptr
0000000000000000 T init_module
                 U msleep
                 U owner
0000000000000000 T parrot_open
                 U parrot_open_ptr
                 U parrot_owner_ptr
0000000000000000 T parrot_read
                 U parrot_read_ptr
0000000000000000 T parrot_release
                 U parrot_release_ptr
                 U printk
0000000000000000 T rust_begin_unwind
0000000000000000 T rust_eh_personality
0000000000000000 T rust_eh_unwind_resume
                 U unregister_chrdev_region
0000000000000000 b knock_out::GLOBAL_STATE::h798a7420bf5e1280
0000000000000000 b _ZN9knock_out13FRAME_COUNTER17h102e5bb3cce2fc0cE.0.0
```
where any `U` symbol comes from a Linux kernel function, and that's nice.

But when we want to use `core::fmt::write` in parrot, the symbol table is:

```bash
                 U alloc_chrdev_region
                 U cdev_add
                 U cdev_del
                 U cdev_init
                 U cdev_ptr
0000000000000140 T cleanup_module
                 U copy_to_user_ffi
                 U fops_ptr
0000000000000000 r GCC_except_table11
0000000000000000 T init_module
                 U msleep
                 U owner
0000000000000000 T parrot_open
                 U parrot_open_ptr
                 U parrot_owner_ptr
0000000000000000 T parrot_read
                 U parrot_read_ptr
0000000000000000 T parrot_release
                 U parrot_release_ptr
                 U printk
0000000000000000 T rust_begin_unwind
0000000000000000 T rust_eh_personality
0000000000000000 T rust_eh_unwind_resume
                 U unregister_chrdev_region
                 U _$LT$str$u20$as$u20$core..fmt..Display$GT$::fmt::h022d13e8e852797c
0000000000000000 t _$LT$$RF$T$u20$as$u20$core..fmt..Display$GT$::fmt::h0fc9a21f3f453b05
                 U core::fmt::write::hf0f4d8fbaa69430d
                 U core::fmt::Formatter::write_fmt::h897518038e53bf5a
0000000000000000 t core::ptr::real_drop_in_place::had8bc5225b63e9eb
0000000000000000 t core::result::unwrap_failed::h6d55a6d53acd12ab
                 U core::panicking::panic_fmt::h58293202aee40982
0000000000000000 t _$LT$$RF$mut$u20$W$u20$as$u20$core..fmt..Write$GT$::write_char::h49b9b11c4df16eec
0000000000000000 t _$LT$$RF$mut$u20$W$u20$as$u20$core..fmt..Write$GT$::write_fmt::h864bf3d1f354c700
0000000000000000 t _$LT$$RF$mut$u20$W$u20$as$u20$core..fmt..Write$GT$::write_str::h7398b4fd9d81b7dd
                 U _$LT$core..fmt..Error$u20$as$u20$core..fmt..Debug$GT$::fmt::h697daa12e9a8bbf2
0000000000000000 t _$LT$parrot..ParrorError$u20$as$u20$core..fmt..Display$GT$::fmt::h915607a16c15eeb8
0000000000000000 b parrot::KERNEL_LOG::he56c6d454f44fcb5
0000000000000000 b parrot::GLOBAL_STATE::h64c1d6ec637cc67d
0000000000000000 b _ZN6parrot13FRAME_COUNTER17hb914cc86267de4e4E.0.0
```

We can see that functions comming from this modules are marked `U`, e.g. `core::fmt::write::hf0f4d8fbaa69430d`, that means they are located on another object file.

### Workaround 1
To handle this problem, I've decided to let `cargo` generate a standard staticlib (instead of `--emit obj=...`. So I've slightly changed `Cargo.toml` by adding:

```toml
[lib]
name = "parrot"
crate-type = ["staticlib"]
```
and changed `Makefile`:

```Makefile
obj-m += parrot.o
parrot-objs := src/ffi.o libparrot.a
ldflags-y += --undefined=init_module --undefined=cleanup_module

all modules:
	make -C /lib/modules/$(shell uname -r)/build M=$(PWD) modules V=1

$(obj)/libparrot.a: $(src)/src/lib.rs $(src)/Cargo.toml
	cd $(src) && cargo rustc --release --lib -- -C relocation-model=static -C code-model=kernel -Z plt=y
	cp $(src)/target/release/libparrot.a $(obj)

clean:
	cargo clean
	make -C /lib/modules/$(shell uname -r)/build M=$(PWD) clean
```

after consulting [kmod](https://github.com/saschagrunert/kmod).

And that builds, a `parrot.ko` is created, this change can even apply to the current `parrot` and generates an identical kernel module. 

### Problem 2
Unfortunately, it doesn't work on my forked branch. I've discovered that, using `dmesg`, the kernel rejects modules which contains `R_X86_64_GOTPCREL` in its relocation table, and the new `parrot.ko` contains a lot:

```bash
readelf -sr parrot.ko | grep R_X86_64_GOTPCREL

...
000000000045  0a5700000009 R_X86_64_GOTPCREL 0000000000000000 _ZN4core3fmt3num53_$LT - 4
00000000009a  09e100000009 R_X86_64_GOTPCREL 0000000000000000 _ZN4core3fmt5write17hf - 4
...
```

 While`-Z plt=y` is a method to avoid `R_X86_64_GOTPCREL` for `parrot`, it cannot prevent the existing the sysroot `core` crates contains such a relocation type (is this true?).

### Workaround 2
A method to handle this problem is to recompile `core` crates using a [custom](https://github.com/rust-osdev/cargo-xbuild) target configuration, and I've managed to do that in another toy kernel module.

Are you interesting in reviewing a such a PR or it's not worth to expand `parrot` in this direction?

Thank you.

**Comment by jbaublitz** [comment]
Hello, thank you for the detailed explanation of your pull request. I'll address both problems individually. This makes a lot of sense from an education perspective but I do want to discuss before merging your PR.

### Problem 1
I've reproduced the scenario you mentioned and compiling as a static library will unfortunately add the entire core library into the executable. Given that this is for education on how to do this, I do not want someone to replicate including all of core statically and see this as the best way to achieve this. There are two concerns I have with this. One is licensing - if core is statically linked to GPL code, I'd have to do more research into the compatibility of the licenses. The other concern is code bloat. This would include the entire core library into the kernel module for every Rust kernel module which I don't think is the way to go. Either unused symbols should be stripped or I have proposed another solution that ties into problem 2.

### Problem 2
This brings up another good point. Cross compiling core definitely works as one solution. Ideally I'd like to do the following as long as licensing permits it: make this project what it was originally intended to be: an interface easing kernel development in Rust. I'd like to do something like exposing `core` to any kernel module that needs the library so it isn't statically compiled into each Rust module. This is something I'd be happy to collaborate on with you. This would provide core in a reusable way as opposed to statically linked into every Rust kernel module. I'd have to do more research into the mechanics of how this could be done, but I'm certainly interested in looking into this.

What are your thoughts?

**Comment by tathanhdinh** [comment]
Hi @jbaublitz 

Thank you for the feedback. I really appreciate it.

> The other concern is code bloat. This would include the entire core library into the kernel module for every Rust kernel module which I don't think is the way to go.

Absolutely agree with you, I didn't think about the fact that all the `core` will be linked into the module. The I was wrong to say that this method give an identical kernel module (with yours using `--emit obj=...`)

> I'd like to do something like exposing core to any kernel module that needs the library so it isn't statically compiled into each Rust module.

Yeah, that would be a very interesting idea, and

> This is something I'd be happy to collaborate on with you

I'm more than happy to hear that and willing to collaborate too. The only worry here is I may not on the same "level" yet: I"m just starting to learn about writing kernel modules with Rust (is this acceptable for you?) .

Thank again for you feedback. It encourages me a lot.

**Comment by jbaublitz** [comment]
> I'm more than happy to hear that and willing to collaborate too. The only worry here is I may not on the same "level" yet: I"m just starting to learn about writing kernel modules with Rust (is this acceptable for you?) .

Let's give it a shot then. The only way to get better is to practice. This will be new territory for me too so we'll both be learning.

My guess as to how this will work is that we will have to both incorporate core into a module and do some symbol resolution whenever a Rust kernel module is loaded. If I recall correctly, `insmod` does some work resolving symbols itself so I'm going to start there and look into how that works. I'll divide out some things that I'd love your help on once I figure out the plan of action.

**Comment by jbaublitz** [comment]
@tathanhdinh I did some research and unfortunately generics can only be supported through compiling individually with each crate. I want to give this some more thought but putting `libcore` in a separate module won't work.

**Comment by jbaublitz** [comment]
@tathanhdinh I'm going to try to dig into this more at some point but I'm not entirely sure what the best way forward is. When I compile it as a static lib, it fails to load in the kernel but contains `std` symbols. When I emit it as an object file, it loads in the kernel but doesn't contain `std` symbols. If you get a chance to take a look at this, let me know if you discover any solution that solves both problems.

**Comment by tathanhdinh** [comment]
Hi @jbaublitz , and very sorry for a long delay.

Sure, I will came back soon to the problem.

**Comment by jbaublitz** [comment]
I've done some reworking of the kernel module and I've fixed the reported problem of undefined symbols from libcore. I'll push the changes shortly and close this out.

**Comment by jbaublitz** [comment]
Closed via b70d8c6b5d4bc52030ff7ae4e58fd4ec66b745c9.
